### PR TITLE
GRAM & RES: improve attributes grammar

### LINKED ISSUE #2362: Incorrect error on procedural macros with colons (`::`) in their names
When a procedural macro has a `::` in its name, a syntax error is shown even thought the code is prefectly valid rust.

Example:

```rust
#![feature(proc_macro, specialization)]

extern crate pyo3;

use pyo3::prelude::*;

#[py::class]
struct MyClass {
    num: i32,
    debug: bool,
    token: PyToken,
}
```

Cargo.toml:

```toml
[package]
authors = ["konstin"]
name = "mylib"
version = "0.1.0"
crate-type = ["cdylib"]

[dependencies.pyo3]
version = "0.2.5"
features = ["extension-module"]
```

The following incorrect error is shown in line 7: `'!' expected, got '::'`

### PR DESCRIPTION
**Pull request by Undin**
* Support any tokens in attributes
* Implement `ContributedReferenceHost` for `RsMetaItem` to inject custom references into attributes
* Implement trait reference resolution in `#[derive()]` attribute via `PsiReferenceProvider`

Closes #2362

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/core/psi/ext/RsReferenceElement.kt [hunk 1]
```diff
@@ -8,11 +8,20 @@ package org.rust.lang.core.psi.ext
 import com.intellij.psi.PsiElement
 import org.rust.lang.core.resolve.ref.RsReference
 
-interface RsReferenceElement : RsElement {
+interface RsWeekReferenceElement : RsElement {
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/psi/ext/RsReferenceElement.kt]: week?..

File: src/main/kotlin/org/rust/lang/core/resolve/ref/RsReferenceBase.kt [hunk 2]
```diff
@@ -28,18 +28,27 @@ abstract class RsReferenceBase<T : RsReferenceElement>(
...
     final override fun getRangeInElement(): TextRange = super.getRangeInElement()
 
     final override fun calculateDefaultRangeInElement(): TextRange {
-        val anchor = element.referenceAnchor
-        check(anchor.parent === element)
-        return TextRange.from(anchor.startOffsetInParent, anchor.textLength)
+        val anchor = element.referenceAnchor ?: return TextRange.EMPTY_RANGE
+        var offset = anchor.startOffsetInParent
+        var parent = anchor.parent
+        while (parent != null && parent != element) {
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/resolve/ref/RsReferenceBase.kt]: Looks like we don't need it now.

**Comment by Undin** [comment]
bors r=vlad20012

**Pull request merged**